pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+
+	// BatchSize specifies the maximum number of messages that can go in a batch
+	// for sending.
+	BatchSize int
+}
+
+// SendDelayDefault is the value for TopicOptions.SendDelay if it is not set.
+const SendDelayDefault = time.Millisecond
+
+// Send publishes a message. It only returns after the message has been
+// sent, or failed to be sent. Send can be called from multiple goroutines
+// at once.
+func (t *Topic) Send(ctx context.Context, m *Message) error {
+	mec := msgErrChan{
+		msg:     m,
+		errChan: make(chan error),
+	}
+	size := len(m.Body)
+	for k, v := range m.Attributes {
+		size += len(k)
+		size += len(v)
+	}
+	if err := t.batcher.AddWait(ctx, mec, size); err != nil {
+		return err
+	}
+	return <-mec.errChan
+}
+
+// Close flushes pending message sends and disconnects the Topic.
+// It only returns after all pending messages have been sent.
+func (t *Topic) Close() error {
+	t.batcher.Flush()
+	return t.driver.Close()
+}
+
+// NewTopic makes a pubsub.Topic from a driver.Topic and opts to
+// tune how messages are sent. Behind the scenes, NewTopic spins up a goroutine
+// to bundle messages into batches and send them to the server.
+func NewTopic(ctx context.Context, d driver.Topic, opts TopicOptions) *Topic {
+	handler := func(item interface{}) {
+		mecs, ok := item.([]msgErrChan)
+		if !ok {
+			panic("failed conversion to []msgErrChan in bundler handler")
+		}
+		var dms []*driver.Message
+		for _, mec := range mecs {
+			m := mec.msg
+			dm := &driver.Message{
+				Body:       m.Body,
+				Attributes: m.Attributes,
+				AckID:      m.ackID,
+			}
+			dms = append(dms, dm)
+		}
+		err := d.SendBatch(ctx, dms)
+		for _, mec := range mecs {
+			mec.errChan <- err
+		}
+	}
+	b := bundler.NewBundler(msgErrChan{}, handler)
+	if opts.SendDelay == 0 {
+		b.DelayThreshold = SendDelayDefault
+	} else {
+		b.DelayThreshold = opts.SendDelay
+	}
+	if opts.BatchSize != 0 {
+		b.BundleCountThreshold = opts.BatchSize
+	}
+	t := &Topic{
+		driver:  d,
+		batcher: b,
+	}
+	return t
+}
+
+// Subscription receives published messages.
+type Subscription struct {
+	driver driver.Subscription
+
+	// ackBatcher makes batches of acks and sends them to the server.
+	ackBatcher *bundler.Bundler
+
+	// q is the local queue of messages downloaded from the server.
+	sem chan struct{}
+	q   []*Message
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
+	// AckDelay tells the max duration to wait before sending the next batch
+	// of acknowledgements back to the server.
+	AckDelay time.Duration
+
+	// AckBatchSize is the maximum number of acks that should be sent to
+	// the server in a batch.
+	AckBatchSize int
+
+	// AckDeadline tells how long the server should wait before assuming a
+	// received message has failed to be processed.
+	AckDeadline time.Duration
+}
+
+// AckDelayDefault is the value for SubscriptionOptions.AckDelay if it is not set.
+const AckDelayDefault = time.Millisecond
```

Comment:
Same comment here as above for SendDelay.

Reply:
Same response. :-)